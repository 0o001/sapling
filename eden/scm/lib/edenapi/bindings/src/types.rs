/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! These types have a representation that is visible to C / C++, and are either
//! allocated in C / C++ and must be freed accordingly, or do not need to be freed
//! at all (no heap allocations).

use anyhow::Error;
use anyhow::Result;
use edenapi_types::metadata::ContentId as ApiContentId;
use edenapi_types::metadata::Sha1 as ApiSha1;
use edenapi_types::metadata::Sha256 as ApiSha256;
use edenapi_types::FileType as ApiFileType;
use edenapi_types::TreeAttributes as ApiTreeAttributes;
use libc::size_t;
use types::HgId as ApiHgId;
use types::Key as ApiKey;
use types::Parents as ApiParents;
use types::RepoPathBuf;

use crate::ptr_len_to_slice;

#[repr(C)]
pub struct Key {
    path: *const u8,
    path_len: size_t,
    hgid: [u8; 20],
}

// Conversion takes a reference to Key, leaving the caller
// responsible for deallocation of the string.
impl TryFrom<&Key> for ApiKey {
    type Error = Error;
    fn try_from(v: &Key) -> Result<Self> {
        let path = unsafe { ptr_len_to_slice(v.path, v.path_len)? };
        // to_vec copies the slice, so the ApiKey can be
        // deallocated in Rust-land.
        let path = RepoPathBuf::from_utf8(path.to_vec())?;
        Ok(Self {
            path,
            hgid: ApiHgId::from_byte_array(v.hgid),
        })
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub enum FileType {
    Regular,
    Executable,
    Symlink,
}

impl From<ApiFileType> for FileType {
    fn from(v: ApiFileType) -> Self {
        match v {
            ApiFileType::Regular => FileType::Regular,
            ApiFileType::Executable => FileType::Executable,
            ApiFileType::Symlink => FileType::Symlink,
        }
    }
}

#[repr(C, u8)]
#[derive(Copy, Clone)]
pub enum Parents {
    None,
    One([u8; 20]),
    Two([u8; 20], [u8; 20]),
}

impl From<ApiParents> for Parents {
    fn from(v: ApiParents) -> Self {
        match v {
            ApiParents::None => Parents::None,
            ApiParents::One(id1) => Parents::One(id1.into_byte_array()),
            ApiParents::Two(id1, id2) => Parents::Two(id1.into_byte_array(), id2.into_byte_array()),
        }
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct HgId(pub [u8; 20]);

impl From<ApiHgId> for HgId {
    fn from(v: ApiHgId) -> Self {
        HgId(v.into_byte_array())
    }
}

impl From<[u8; 20]> for HgId {
    fn from(v: [u8; 20]) -> Self {
        HgId(v)
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct ContentId(pub [u8; 32]);

impl From<ApiContentId> for ContentId {
    fn from(v: ApiContentId) -> Self {
        ContentId(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha1(pub [u8; 20]);

impl From<ApiSha1> for Sha1 {
    fn from(v: ApiSha1) -> Self {
        Sha1(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct Sha256(pub [u8; 32]);

impl From<ApiSha256> for Sha256 {
    fn from(v: ApiSha256) -> Self {
        Sha256(v.into())
    }
}

#[repr(C)]
#[derive(Copy, Clone)]
pub struct TreeAttributes {
    manifest_blob: bool,
    parents: bool,
    child_metadata: bool,
}

impl From<TreeAttributes> for ApiTreeAttributes {
    fn from(v: TreeAttributes) -> Self {
        ApiTreeAttributes {
            manifest_blob: v.manifest_blob,
            parents: v.parents,
            child_metadata: v.child_metadata,
        }
    }
}
