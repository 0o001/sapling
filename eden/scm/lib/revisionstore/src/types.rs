/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use anyhow::Result;
use bytes::Bytes;
use serde_derive::{Deserialize, Serialize};
use sha2::Digest;

#[cfg(any(test, feature = "for-tests"))]
use rand::Rng;

use types::{Key, Sha256};

/// Kind of content hash stored in the LFS pointer. Adding new types is acceptable, re-ordering or
/// removal is forbidden.
#[derive(
    Serialize,
    Deserialize,
    PartialEq,
    Eq,
    Debug,
    Hash,
    Ord,
    PartialOrd,
    Clone
)]
pub enum ContentHash {
    Sha256(Sha256),
}

#[derive(Clone, PartialEq, Eq, Debug, Hash, Ord, PartialOrd)]
pub enum StoreKey {
    HgId(Key),
    /// The Key is a temporary workaround to being able to fallback from the LFS protocol to the
    /// non-LFS protocol. Do not depend on it as it will be removed.
    Content(ContentHash, Option<Key>),
}

impl ContentHash {
    pub fn sha256(data: &Bytes) -> Result<Self> {
        let mut hash = sha2::Sha256::new();
        hash.input(data);

        let bytes: [u8; Sha256::len()] = hash.result().into();
        Ok(ContentHash::Sha256(Sha256::from_slice(&bytes)?))
    }

    pub fn unwrap_sha256(self) -> Sha256 {
        match self {
            ContentHash::Sha256(hash) => hash,
        }
    }
}

impl StoreKey {
    pub fn hgid(key: Key) -> Self {
        StoreKey::HgId(key)
    }

    pub fn content(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }
}

impl From<Key> for StoreKey {
    fn from(k: Key) -> Self {
        StoreKey::HgId(k)
    }
}

impl<'a> From<&'a Key> for StoreKey {
    fn from(k: &'a Key) -> Self {
        StoreKey::HgId(k.clone())
    }
}

impl From<ContentHash> for StoreKey {
    fn from(hash: ContentHash) -> Self {
        StoreKey::Content(hash, None)
    }
}

impl<'a> From<&'a ContentHash> for StoreKey {
    fn from(hash: &'a ContentHash) -> Self {
        StoreKey::Content(hash.clone(), None)
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl quickcheck::Arbitrary for ContentHash {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        ContentHash::Sha256(Sha256::arbitrary(g))
    }
}

#[cfg(any(test, feature = "for-tests"))]
impl quickcheck::Arbitrary for StoreKey {
    fn arbitrary<G: quickcheck::Gen>(g: &mut G) -> Self {
        if g.gen() {
            StoreKey::HgId(Key::arbitrary(g))
        } else {
            StoreKey::Content(ContentHash::arbitrary(g), Option::arbitrary(g))
        }
    }
}
