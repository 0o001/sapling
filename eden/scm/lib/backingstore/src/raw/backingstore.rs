/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Provides the c-bindings for `crate::backingstore`.

use std::slice;
use std::str;

use anyhow::ensure;
use anyhow::Error;
use anyhow::Result;
use libc::c_char;
use libc::c_void;
use libc::size_t;
use manifest::List;
use revisionstore::scmstore::FileAuxData as ScmStoreFileAuxData;
use types::Key;

use crate::backingstore::BackingStore;
use crate::raw::CBytes;
use crate::raw::CFallible;
use crate::raw::FileAuxData;
use crate::raw::Request;
use crate::raw::Tree;

fn stringpiece_to_slice<'a, T, U>(ptr: *const T, length: size_t) -> Result<&'a [U]> {
    ensure!(!ptr.is_null(), "string ptr is null");
    Ok(unsafe { slice::from_raw_parts(ptr as *const U, length) })
}

fn backingstore_new(
    repository: *const c_char,
    repository_len: size_t,
    use_edenapi: bool,
    aux_data: bool,
) -> Result<*mut BackingStore> {
    super::init::backingstore_global_init();

    let repository = stringpiece_to_slice(repository, repository_len)?;
    let repo = str::from_utf8(repository)?;
    let store = Box::new(BackingStore::new(repo, use_edenapi, aux_data)?);

    Ok(Box::into_raw(store))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_new(
    repository: *const c_char,
    repository_len: size_t,
    use_edenapi: bool,
    aux_data: bool,
) -> CFallible<BackingStore> {
    backingstore_new(repository, repository_len, use_edenapi, aux_data).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_free(store: *mut BackingStore) {
    assert!(!store.is_null());
    let store = unsafe { Box::from_raw(store) };
    drop(store);
}

fn backingstore_get_blob(
    store: *mut BackingStore,
    name: *const u8,
    name_len: usize,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut CBytes> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let path = stringpiece_to_slice(name, name_len)?;
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_blob(path, node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no blob found")))
        .map(CBytes::from_vec)
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_blob(
    store: *mut BackingStore,
    name: *const u8,
    name_len: usize,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<CBytes> {
    backingstore_get_blob(store, name, name_len, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_blob_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<CBytes>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_blob_batch(keys, local, |idx, result| {
        let result = result
            .and_then(|opt| opt.ok_or_else(|| Error::msg("no blob found")))
            .map(CBytes::from_vec)
            .map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

fn backingstore_get_tree(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut Tree> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_tree(node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no tree found")))
        .and_then(|list| list.try_into())
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_tree(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<Tree> {
    backingstore_get_tree(store, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_tree_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<Tree>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_tree_batch(keys, local, |idx, result| {
        let result: Result<List> =
            result.and_then(|opt| opt.ok_or_else(|| Error::msg("no tree found")));
        let result: Result<Tree> = result.and_then(|list| list.try_into());
        let result: Result<*mut Tree> = result.map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

fn backingstore_get_file_aux(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> Result<*mut FileAuxData> {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let node = stringpiece_to_slice(node, node_len)?;

    store
        .get_file_aux(node, local)
        .and_then(|opt| opt.ok_or_else(|| Error::msg("no file aux data found")))
        .map(|aux| aux.into())
        .map(|result| Box::into_raw(Box::new(result)))
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_file_aux(
    store: *mut BackingStore,
    node: *const u8,
    node_len: usize,
    local: bool,
) -> CFallible<FileAuxData> {
    backingstore_get_file_aux(store, node, node_len, local).into()
}

#[no_mangle]
pub extern "C" fn rust_backingstore_get_file_aux_batch(
    store: *mut BackingStore,
    requests: *const Request,
    size: usize,
    local: bool,
    data: *mut c_void,
    resolve: unsafe extern "C" fn(*mut c_void, usize, CFallible<FileAuxData>),
) {
    assert!(!store.is_null());
    let store = unsafe { &*store };
    let requests: &[Request] = unsafe { slice::from_raw_parts(requests, size) };
    let keys: Vec<Result<Key>> = requests.iter().map(|req| req.try_into_key()).collect();

    store.get_file_aux_batch(keys, local, |idx, result| {
        let result: Result<ScmStoreFileAuxData> =
            result.and_then(|opt| opt.ok_or_else(|| Error::msg("no file aux data found")));
        let result: Result<FileAuxData> = result.map(|aux| aux.into());
        let result: Result<*mut FileAuxData> = result.map(|result| Box::into_raw(Box::new(result)));
        unsafe {
            resolve(data, idx, result.into())
        };
    });
}

#[no_mangle]
pub extern "C" fn rust_tree_free(tree: *mut Tree) {
    assert!(!tree.is_null());
    let tree = unsafe { Box::from_raw(tree) };
    drop(tree);
}

#[no_mangle]
pub extern "C" fn rust_file_aux_free(aux: *mut FileAuxData) {
    assert!(!aux.is_null());
    let aux = unsafe { Box::from_raw(aux) };
    drop(aux);
}

#[no_mangle]
pub extern "C" fn rust_backingstore_flush(store: *mut BackingStore) {
    assert!(!store.is_null());
    let store = unsafe { &*store };

    store.flush();
}
