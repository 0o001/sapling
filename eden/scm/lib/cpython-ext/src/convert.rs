/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

use cpython::*;
use serde::Deserialize;
use serde::Serialize;
use std::cmp::Ord;
use std::cmp::Ordering;
use std::cmp::PartialOrd;
use std::hash::Hash;
use std::hash::Hasher;
use std::ops::Deref;

/// Wrapper type. Converts between pure Rust bytes-like types and PyBytes.
///
/// The Rust type needs to implement `AsRef<[u8]>` and `From<Vec<u8>>`.
///
/// In bindings code:
/// - For input, use `v: BytesLike<MyType>` in definition, and `v.0` to extract
///   `MyType`.
/// - For output, use `-> BytesLike<MyType>` in definition, and `BytesLike(v)`
///   to construct the return value.
#[derive(Clone, Debug, Eq, Ord, Hash, PartialEq, PartialOrd)]
pub struct BytesLike<T>(pub T);

impl<T: AsRef<[u8]>> ToPyObject for BytesLike<T> {
    type ObjectType = PyBytes;

    fn to_py_object(&self, py: Python) -> Self::ObjectType {
        PyBytes::new(py, self.0.as_ref())
    }
}

impl<'s, T: From<Vec<u8>>> FromPyObject<'s> for BytesLike<T> {
    fn extract(py: Python, obj: &'s PyObject) -> PyResult<Self> {
        obj.extract::<PyBytes>(py)
            .map(|v| Self(v.data(py).to_vec().into()))
    }
}

impl<T> Deref for BytesLike<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

/// Wrapper type. Converts between pure Rust serde types and PyObjct.
///
/// In bindings code:
/// - For input, use `v: Serde<MyType>` in definition, and `v.0` to extract
///   `MyType`.
/// - For output, use `-> Serde<MyType>` in definition, and `Serde(v)` to
///   construct the return value.
#[derive(Debug)]
pub struct Serde<T>(pub T);

impl<T: Serialize> ToPyObject for Serde<T> {
    type ObjectType = PyObject;

    fn to_py_object(&self, py: Python) -> Self::ObjectType {
        crate::ser::to_object(py, &self.0).unwrap()
    }
}

impl<'s, T> FromPyObject<'s> for Serde<T>
where
    T: for<'de> Deserialize<'de>,
{
    fn extract(py: Python, obj: &'s PyObject) -> PyResult<Self> {
        let inner = crate::de::from_object(py, obj.clone_ref(py))?;
        Ok(Self(inner))
    }
}

impl<T> Deref for Serde<T> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<T: PartialOrd> PartialOrd<Serde<T>> for Serde<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.0.partial_cmp(&other.0)
    }
}

impl<T: Ord> Ord for Serde<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0.cmp(&other.0)
    }
}

impl<T: Hash> Hash for Serde<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state)
    }
}

impl<T: PartialEq> PartialEq for Serde<T> {
    fn eq(&self, other: &Self) -> bool {
        self.0.eq(&other.0)
    }
}

impl<T: Eq> Eq for Serde<T> {}

impl<T: Clone> Clone for Serde<T> {
    fn clone(&self) -> Self {
        Self(self.0.clone())
    }
}
