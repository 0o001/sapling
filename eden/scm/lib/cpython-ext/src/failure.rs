/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Integrate cpython with failure

use cpython::{ObjectProtocol, PyClone, PyResult, Python, PythonObjectWithTypeObject};
use failure::{AsFail, Error, Fallible};
use std::fmt;

/// Extends the `Result` type to allow conversion to `PyResult` by specifying a
/// Python Exception class to convert to.
///
/// If the error is created via [`FallibleExt`], the original Python error
/// will be returned.
///
/// # Examples
///
/// ```
/// use cpython::{exc, Python, PyResult};
/// use cpython_ext::failure::ResultPyErrExt;
/// use failure::{format_err, Error};
///
/// fn fail_if_negative(i: i32) -> Result<i32, Error> {
///    if (i >= 0) {
///       Ok(i)
///    } else {
///       Err(format_err!("{} is negative", i))
///    }
/// }
///
/// fn py_fail_if_negative(py: Python, i: i32) -> PyResult<i32> {
///    fail_if_negative(i).map_pyerr::<exc::ValueError>(py)
/// }
/// ```
pub trait ResultPyErrExt<T> {
    fn map_pyerr<PE: PythonObjectWithTypeObject>(self, py: Python<'_>) -> PyResult<T>;
}

/// Extends the `PyResult` type to allow conversion to `Result`.
///
/// # Examples
///
/// ```
/// use cpython_ext::failure::{FallibleExt, PyErr};
/// use failure::Fallible;
///
/// fn eval_py() -> Fallible<i32> {
///     let gil = cpython::Python::acquire_gil();
///     let py = gil.python();
///     let obj = py.eval("1 + 2", None, None).into_fallible()?;
///     obj.extract(py).into_fallible()
/// }
///
/// fn round_trip() -> cpython::PyResult<()> {
///     let gil = cpython::Python::acquire_gil();
///     let py = gil.python();
///     let res = py.eval("1 + 2", None, None).into_fallible();
///     use cpython_ext::failure::ResultPyErrExt;
///     res.map(|_| ()).map_pyerr::<cpython::exc::ValueError>(py)
/// }
/// ```
///
pub trait FallibleExt<T> {
    fn into_fallible(self) -> Fallible<T>;
}

impl<T, E: AsFail> ResultPyErrExt<T> for Result<T, E> {
    fn map_pyerr<PE: PythonObjectWithTypeObject>(self, py: Python<'_>) -> PyResult<T> {
        self.map_err(|e| {
            let e = e.as_fail();
            if let Some(e) = e.downcast_ref::<PyErr>() {
                e.inner.clone_ref(py)
            } else {
                cpython::PyErr::new::<PE, _>(py, e.to_string())
            }
        })
    }
}

impl<T> FallibleExt<T> for PyResult<T> {
    fn into_fallible(self) -> Fallible<T> {
        self.map_err(|e| Error::from_boxed_compat(Box::new(PyErr::from(e))))
    }
}

/// An enhanced version of `PyErr` that implements [`std::error::Error`].
pub struct PyErr {
    inner: cpython::PyErr,
}

impl From<cpython::PyErr> for PyErr {
    fn from(e: cpython::PyErr) -> PyErr {
        PyErr { inner: e }
    }
}

impl From<PyErr> for cpython::PyErr {
    fn from(e: PyErr) -> cpython::PyErr {
        e.inner
    }
}

impl fmt::Display for PyErr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let repr = self
            .inner
            .pvalue
            .as_ref()
            .unwrap_or_else(|| &self.inner.ptype)
            .repr(py)
            .map(|s| s.to_string_lossy(py).to_string())
            .unwrap_or_else(|_| "<error in repr>".into());
        write!(f, "{}", repr)
    }
}

impl fmt::Debug for PyErr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self)
    }
}

impl std::error::Error for PyErr {}

impl serde::ser::Error for PyErr {
    fn custom<T: std::fmt::Display>(msg: T) -> Self {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let err = cpython::PyErr::new::<cpython::exc::TypeError, _>(py, msg.to_string());
        Self { inner: err }
    }
}

impl PyErr {
    pub fn into_inner(self) -> cpython::PyErr {
        self.inner
    }
}
