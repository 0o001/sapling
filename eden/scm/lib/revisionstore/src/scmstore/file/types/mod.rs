/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod auxdata;
mod lazy_file;
mod store_file;

pub(crate) use self::lazy_file::LazyFile;
pub use self::{attrs::FileAttributes, auxdata::FileAuxData, store_file::StoreFile};
