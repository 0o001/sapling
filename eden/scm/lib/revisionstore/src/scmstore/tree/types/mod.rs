/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

mod attrs;
mod lazy_tree;
mod store_tree;

pub(crate) use self::lazy_tree::LazyTree;
pub use self::{attrs::TreeAttributes, store_tree::StoreTree};
