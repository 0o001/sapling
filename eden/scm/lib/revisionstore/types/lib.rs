/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

#![deny(warnings)]

//! revisionstore_types - Data types used by the revisionstore crate

mod datastore;

pub use crate::datastore::Metadata;
