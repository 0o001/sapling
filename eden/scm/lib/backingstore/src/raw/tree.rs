/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This software may be used and distributed according to the terms of the
 * GNU General Public License version 2.
 */

//! Representation of tree in EdenFS.
//!
//! Structs in this file should be keep in sync with `eden/fs/model/{Tree, TreeEntry}.h`.

use crate::raw::CBytes;
use failure::{format_err, Fallible};
use manifest::{tree::List, FileType, FsNode};
use std::convert::TryFrom;
use types::PathComponentBuf;

#[repr(u8)]
pub enum TreeEntryType {
    Tree,
    RegularFile,
    ExecutableFile,
    Symlink,
}

impl From<FileType> for TreeEntryType {
    fn from(file_type: FileType) -> Self {
        match file_type {
            FileType::Regular => TreeEntryType::RegularFile,
            FileType::Executable => TreeEntryType::ExecutableFile,
            FileType::Symlink => TreeEntryType::Symlink,
        }
    }
}

#[repr(C)]
pub struct TreeEntry {
    hash: CBytes,
    name: CBytes,
    ttype: TreeEntryType,
    // Using pointer as `Option<T>`
    size: *mut u64,
    content_sha1: *mut CBytes,
}

impl TreeEntry {
    fn try_from_path_node(path: PathComponentBuf, node: FsNode) -> Fallible<Self> {
        let (ttype, hash) = match node {
            FsNode::Directory(Some(hgid)) => (TreeEntryType::Tree, hgid.as_ref().to_vec()),
            FsNode::File(metadata) => (metadata.file_type.into(), metadata.hgid.as_ref().to_vec()),
            _ => return Err(format_err!("received an ephemeral directory")),
        };

        Ok(TreeEntry {
            hash: hash.into(),
            name: path.as_ref().as_byte_slice().to_vec().into(),
            ttype,
            // TODO: we currently do not have these information stored in Mercurial.
            size: std::ptr::null_mut(),
            content_sha1: std::ptr::null_mut(),
        })
    }
}

#[repr(C)]
pub struct Tree {
    entries: *const TreeEntry,
    /// This makes sure `entries` above is pointing to a valid memory.
    entries_ptr: *mut Vec<TreeEntry>,
    length: usize,
    hash: CBytes,
}

impl TryFrom<List> for Tree {
    type Error = failure::Error;

    fn try_from(list: List) -> Result<Self, Self::Error> {
        match list {
            List::NotFound | List::File => Err(format_err!("not found")),
            List::Directory(list) => {
                let entries = list
                    .into_iter()
                    .map(|(path, node)| TreeEntry::try_from_path_node(path, node))
                    .collect::<Fallible<Vec<_>>>()?;

                let entries = Box::new(entries);
                let length = entries.len();

                Ok(Tree {
                    entries: entries.as_ptr(),
                    entries_ptr: Box::into_raw(entries),
                    length,
                    hash: Vec::new().into(),
                })
            }
        }
    }
}

impl Drop for Tree {
    fn drop(&mut self) {
        let entry = unsafe { Box::from_raw(self.entries_ptr) };
        drop(entry);
    }
}
