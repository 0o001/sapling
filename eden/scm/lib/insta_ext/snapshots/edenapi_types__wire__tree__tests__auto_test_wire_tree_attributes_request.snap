---
source: tree.rs
expression: "WireTreeAttributesRequest::arbitrary(&mut g)"

---
{}
